backend/server.js

```diff
@@ -1,30 +1,145 @@
-import mongoose from 'mongoose'
-
-const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/authAPI"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
-mongoose.Promise = Promise
-
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
-const port = process.env.PORT || 8080
-const app = express()
+import express from "express";
+import bodyParser from "body-parser";
+import cors from "cors";
+import mongoose from "mongoose";
+import crypto from "crypto";
+import bcrypt from "bcrypt";
+
+const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/authAPI";
+mongoose.connect(mongoUrl, {
+  useNewUrlParser: true,
+  useUnifiedTopology: true,
+  useCreateIndex: true,
+});
+mongoose.Promise = Promise;
+
+const Thought = mongoose.model("Thought", {
+  message: String,
+});
+
+// at least two properties
+const User = mongoose.model("User", {
+  username: {
+    type: String,
+    required: true,
+    unique: true,
+  },
+  password: {
+    type: String,
+    required: true,
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString("hex"),
+  },
+});
+
+// access token always sent when you send your request frmo front end
+// if user have the access token then next()
+//authenicate middlewear, specific on some end point or always
+const authenticateUser = async (req, res, next) => {
+  const accessToken = req.header("Authorization");
+
+  try {
+    const user = await User.findOne({ accessToken });
+    if (user) {
+      next();
+    } else {
+      res.status(401).json({ success: false, message: "Not authorized" });
+    }
+  } catch (error) {
+    res.status(400).json({ success: false, message: "Invalid request", error });
+  }
+};
+
+// Defines the port the app will run on. Defaults to 8080, but can be
+const port = process.env.PORT || 8080;
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+app.use(cors());
+app.use(bodyParser.json());
 
 // Start defining your routes here
-app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+//GET Requests
+app.get("/", (req, res) => {
+  res.send("Hello world");
+});
+
+//GET Requests to get all messages in Thought model:
+// specify authenticateUser for this end point, do we have a user with this token -> then move on to next()
+app.get("/thoughts", authenticateUser);
+app.get("/thoughts", async (req, res) => {
+  const thoughts = await Thought.find();
+  res.json({ success: true, thoughts });
+});
+
+app.post("/thoughts", authenticateUser);
+app.post("/thoughts", async (req, res) => {
+  const { message } = req.body;
+
+  try {
+    const newThought = await new Thought({ message }).save();
+    res.json({ success: true, newThought });
+  } catch (error) {
+    res.status(400).json({ success: false, message: "Invalid request", error });
+  }
+});
```

Comment:
🧹

Reply:
Hi Karol, even though we are not using these endpoints for the front-end. This is something Maks went through with us for This project, for this particular week. So it is actually not a reused project. So that when the user has an accesstoken, the messages render, that has been sent in the POST /thoughts endpoint.  Could be nice to keep this for the future :)?